### ISSUE TITLE
Enhance `DiceMetric` to support counting negative samples

### ISSUE THREAD

**Issue by yiheng-wang-nv** [Project-MONAI/MONAI#4160]
I think the `DiceMetric` may has an error when there is no foreground for `y` but has foreground for `y_pred`. According to the formula, the result should be 0, but so far the class will return `nan`:

```
torch.where(y_o > 0, (2.0 * intersection) / denominator, torch.tensor(float("nan"), device=y_o.device))
```

The earliest commit I can found is already use this way for calculation: see https://github.com/Project-MONAI/MONAI/pull/285/files#diff-df0f76defe29c2c91286e52334ead7a0f1f54e392d1a3e8280e2e714800dc4cbL96

I think we may need to use:
```
torch.where(denominator > 0, (2.0 * intersection) / denominator, torch.tensor(float("nan"), device=y_o.device))
```
In addition, we use a fixed value: `nan` to fill the cases that the ground truth (as I mentioned, may need to be placed by denominator) has no foreground. However, in some practical situations, people may consider to use other values such as 1 instead. For example, in 
the latest Kaggle competition: https://www.kaggle.com/competitions/uw-madison-gi-tract-image-segmentation/overview/evaluation , it says:
```
where X is the predicted set of pixels and Y is the ground truth. The Dice coefficient is defined to be 1 when both X and Y are empty.
```

Therefore, I think we may need to add an argument here thus users can specify which value to use.

Hi @ristoh @Nic-Ma @wyli @ericspod , could you please help to double check it? Thanks!

**Comment by Nic-Ma** [comment]
Hi @dongyang0122 @holgerroth ,

Could you please help also share some comments on this Dice computation question?

Thanks in advance.

**Comment by dongyang0122** [comment]
There are currently two ways to handle the empty ground truth case.

1. (Ours) Skip calculation and set dice score to NaN;
2. Return one when the prediction is empty, zero otherwise;

We may need to support both since both of them are used by researchers. There is no perfect solution for this situation.

**Comment by ericspod** [comment]
With the dice loss we have [two distinct values for smoothing the numerator and denominator](https://github.com/Project-MONAI/MONAI/blob/d63f4906688bc740067807a7d4e2112f1b006706/monai/losses/dice.py#L179) which can be set to allow the user to select which way to handle empty ground truths.

**Comment by yiheng-wang-nv** [comment]
HI @ericspod , thanks for the suggestion. It seems the smooth based way used in dice loss cannot handle the case of:
```
ground truth is empty, and prediction is not empty
```
Hi @dongyang0122 , could you please help to double check this? Should the way to calculate dice loss (smooth based) and dice coefficient metric (distinguish )be consistent?

**Comment by dongyang0122** [comment]
I do not see the point of having smoothing factors in a dice metric calculation. Even if we can set different smoothing factors to handle empty ground truth, such operation will affect the accuracy of other cases with non-empty ground truth.

**Comment by ericspod** [comment]
If either ground truth or prediction are 0 then the numerator will be 0 but the denominator not, so a metric score of 0. The smooth values are to prevent dividing by 0 if both are empty so should be set extremely small like 1e-5 or 1e-10. The dice loss assumes input values are all positive as it provides sigmoid or softmax activation. So there's a number of differences between the loss and metric which may be an issue if the metric's impression of how well a network is doing diverges a lot from what the loss value would indicate. The PR seems fine to me if that's the expected behaviour but it is very different from how the loss is calculated.

**Comment by dongyang0122** [comment]
In principle, Dice loss is different from Dice metric. Dice loss requires predicted probabilities (floats) from the model in order to be differentiable. It is calculated based on the Dice metric, but it is only an estimate. In contrast, the Dice metric requires segmentation masks after thresholding. Despite their similar names, the underlying logic is completely different.
